File: bridge/service.py
import argparse
from contextlib import closing
import json
import logging
import os
from pathlib import Path
import queue
import re
import signal
import sqlite3
import subprocess
import threading
import time

import paho.mqtt.client as mqtt

from .sources import AGENTS, CodexUsage, PcMetrics, read_state

LOG = logging.getLogger("agentdeck")
IDENTIFIER = re.compile(r"[A-Za-z0-9_-]{1,32}\Z")
ACTIONS = {"confirm", "cancel", "stop", "pause", "resume", "approve", "reject", "retry"}


def load_config(path):
    path = Path(path).resolve()
    cfg = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(cfg.get("host"), str) or not cfg["host"]:
        raise ValueError("host is required")
    if type(cfg.get("port", 1883)) is not int or not 1 <= cfg.get("port", 1883) <= 65535:
        raise ValueError("Invalid MQTT port")
    if not cfg.get("devices") or not all(isinstance(d, str) and IDENTIFIER.fullmatch(d) for d in cfg["devices"]):
        raise ValueError("devices must list authorized device IDs")
    if not 1 <= cfg.get("interval", 3) <= 5:
        raise ValueError("interval must be 1..5 seconds")
    for agent, actions in cfg.get("handlers", {}).items():
        if agent not in AGENTS or not isinstance(actions, dict):
            raise ValueError("Invalid handlers")
        for action, argv in actions.items():
            if action not in ACTIONS or not isinstance(argv, list) or not argv or not all(isinstance(a, str) and a for a in argv):
                raise ValueError("Handlers must be fixed, nonempty argv lists")
    cfg["state_dir"] = str((path.parent / cfg.get("state_dir", ".state")).resolve())
    return cfg


class Bridge:
    def __init__(self, config):
        self.cfg = config
        self.stop = threading.Event()
        self.connected = threading.Event()
        self.commands = queue.Queue(maxsize=16)
        self.outbox = queue.Queue(maxsize=64)
        self.state_dir = Path(config["state_dir"])
        self.state_dir.mkdir(parents=True, exist_ok=True)
        self.metrics = PcMetrics()
        self.usage = CodexUsage(config.get("codex_usage_command"))
        self.client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2,
                                  client_id=config.get("client_id", "agentdeck-pc-main"), clean_session=True)
        user = os.environ.get(config.get("username_env", "AGENTDECK_MQTT_USER"))
        password = os.environ.get(config.get("password_env", "AGENTDECK_MQTT_PASSWORD"))
        if user:
            self.client.username_pw_set(user, password)
        if config.get("tls_ca"):
            self.client.tls_set(ca_certs=config["tls_ca"])
        self.client.reconnect_delay_set(1, 60)
        self.client.max_queued_messages_set(32)
        self.client.will_set("pc/status", json.dumps({"online": False, "ts": int(time.time())}), qos=1, retain=True)
        self.client.on_connect = self._connect
        self.client.on_disconnect = self._disconnect
        self.client.on_message = self._message

    def _connect(self, client, userdata, flags, reason, properties):
        if reason.is_failure:
            LOG.warning("MQTT connection rejected: %s", reason)
            return
        result, _ = client.subscribe([(f"agent/{agent}/command", 1) for agent in AGENTS])
        if result == mqtt.MQTT_ERR_SUCCESS:
            self.connected.set()
            LOG.info("MQTT connected")

    def _disconnect(self, client, userdata, flags, reason, properties):
        self.connected.clear()
        LOG.info("MQTT disconnected")

    def _message(self, client, userdata, message):
        if message.retain or len(message.payload) > 1024:
            return
        parts = message.topic.split("/")
        if len(parts) != 3 or parts[0] != "agent" or parts[1] not in AGENTS or parts[2] != "command":
            return
        try:
            data = json.loads(message.payload)
            if not isinstance(data, dict):
                return
            if data.get("device_id") not in self.cfg["devices"]:
                return
            if not isinstance(data.get("id"), str) or not re.fullmatch(r"[A-Za-z0-9_-]{1,64}", data["id"]):
                return
            self.commands.put_nowait((parts[1], data))
        except (ValueError, UnicodeError, queue.Full):
            LOG.warning("Invalid command or command queue full")

    def ack(self, data, status, reason=""):
        body = {"id": data["id"], "status": status, "reason": reason, "ts": int(time.time())}
        try:
            self.outbox.put_nowait((f"agentdeck/{data['device_id']}/ack", body))
        except queue.Full:
            LOG.warning("Acknowledgement queue full")

    def _worker(self):
        # Persist reservation before side effects: duplicate QoS1 deliveries and bridge
        # restarts cannot execute the same id twice. A crash is reported as unknown.
        with closing(sqlite3.connect(self.state_dir / "commands.sqlite3")) as db:
            db.execute("CREATE TABLE IF NOT EXISTS commands (device TEXT, id TEXT, agent TEXT, action TEXT, status TEXT, ts INTEGER, PRIMARY KEY(device,id))")
            while not self.stop.is_set():
                try:
                    agent, data = self.commands.get(timeout=.2)
                except queue.Empty:
                    continue
                now = int(time.time())
                action = data.get("action")
                if not isinstance(action, str) or action not in ACTIONS or type(data.get("ts")) is not int or type(data.get("expires_at")) is not int or not (
                        now - 15 <= data["ts"] <= now + 5 and now <= data["expires_at"] <= now + 30):
                    self.ack(data, "rejected", "expired_or_invalid"); continue
                key = (data["device_id"], data["id"])
                existing = db.execute("SELECT agent, action, status FROM commands WHERE device=? AND id=?", key).fetchone()
                if existing:
                    same = existing[:2] == (agent, action)
                    self.ack(data, existing[2] if same and existing[2] != "executing" else "rejected", "duplicate_or_unknown")
                    continue
                status, _ = read_state(self.state_dir, agent)
                argv = self.cfg.get("handlers", {}).get(agent, {}).get(action)
                if not status["online"] or not argv:
                    self.ack(data, "rejected", "agent_offline" if not status["online"] else "unsupported_action")
                    continue
                db.execute("INSERT INTO commands VALUES(?,?,?,?,?,?)", (*key, agent, action, "executing", now)); db.commit()
                result = "failed"
                try:
                    # The broker supplies no executable, arguments or shell text.
                    # The trusted handler receives the validated envelope via stdin.
                    completed = subprocess.run(argv, input=json.dumps(dict(data, agent=agent)), text=True,
                                               stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, timeout=5,
                                               creationflags=getattr(subprocess, "CREATE_NO_WINDOW", 0))
                    result = "completed" if completed.returncode == 0 else "failed"
                except (OSError, subprocess.SubprocessError):
                    pass
                db.execute("UPDATE commands SET status=? WHERE device=? AND id=?", (result, *key))
                db.execute("DELETE FROM commands WHERE ts < ?", (now - 86400,)); db.commit()
                self.ack(data, result)

    def publish(self, topic, body, retain=False):
        if self.connected.is_set():
            # Telemetry is QoS0 and never buffered across disconnects. Commands/acks
            # have explicit application ids and expiry instead of hidden retries.
            return self.client.publish(topic, json.dumps(body, ensure_ascii=False, allow_nan=False), qos=0, retain=retain)
        return None

    def sample(self):
        self.publish("pc/status", self.metrics.sample())
        for agent in AGENTS:
            status, usage = read_state(self.state_dir, agent)
            if agent == "codex":
                usage = self.usage.sample() or usage
            self.publish(f"agent/{agent}/status", status)
            self.publish(f"agent/{agent}/usage", dict(usage or {"available": False}, ts=int(time.time())))

    def run(self):
        worker = threading.Thread(target=self._worker, daemon=True)
        worker.start(); self.usage.start()
        self.client.connect_async(self.cfg["host"], self.cfg.get("port", 1883), keepalive=10)
        self.client.loop_start()
        next_sample = 0
        try:
            while not self.stop.wait(.05):
                if time.monotonic() >= next_sample:
                    next_sample = time.monotonic() + self.cfg.get("interval", 3)
                    if self.connected.is_set():
                        try:
                            self.sample()
                        except (OSError, ValueError) as error:
                            LOG.warning("Data source unavailable: %s", type(error).__name__)
                for _ in range(16):
                    try:
                        topic, body = self.outbox.get_nowait()
                    except queue.Empty:
                        break
                    if int(time.time()) - body["ts"] <= 15:
                        self.publish(topic, body)
        finally:
            self.stop.set(); self.usage.stop.set(); worker.join(timeout=6)
            for agent in AGENTS:
                self.publish(f"agent/{agent}/status", {"online": False, "working": False, "ts": int(time.time())})
            sent = self.publish("pc/status", {"online": False, "ts": int(time.time())}, retain=True)
            if sent:
                try:
                    sent.wait_for_publish(timeout=1)
                except RuntimeError:
                    pass
            self.client.disconnect(); self.client.loop_stop()
            if self.usage.thread:
                self.usage.thread.join(timeout=5)


def main():
    parser = argparse.ArgumentParser(description="Agent Deck real MQTT bridge")
    parser.add_argument("--config", default="bridge/config.json")
    args = parser.parse_args()
    logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
    bridge = Bridge(load_config(args.config))
    for sig in (signal.SIGINT, signal.SIGTERM):
        signal.signal(sig, lambda *_: bridge.stop.set())
    bridge.run()


if __name__ == "__main__":
    main()

File: bridge/src/service.rs
use crate::{
    config::{ACTIONS, AGENTS, Config, valid_command_id},
    sources::{PcMetrics, UsageCache},
    state::{read_state, unix_time},
};
use anyhow::Result;
use rumqttc::{
    Client, ConnectionError, Event, Incoming, LastWill, MqttOptions, Outgoing, Publish, QoS,
    Transport,
};
use rusqlite::{Connection, params};
use serde_json::{Value, json};
use std::{
    io::Write,
    process::{Command, Stdio},
    sync::{
        Arc,
        atomic::{AtomicBool, Ordering},
        mpsc,
    },
    thread,
    time::{Duration, Instant},
};

#[derive(Debug)]
struct IncomingCommand {
    agent: String,
    data: Value,
}

fn accepted(publish: &Publish, cfg: &Config) -> Option<IncomingCommand> {
    if publish.retain || publish.payload.len() > 1024 {
        return None;
    }
    let parts: Vec<_> = publish.topic.split('/').collect();
    if parts.len() != 3
        || parts[0] != "agent"
        || !AGENTS.contains(&parts[1])
        || parts[2] != "command"
    {
        return None;
    }
    let data: Value = serde_json::from_slice(&publish.payload).ok()?;
    let object = data.as_object()?;
    let device = object.get("device_id")?.as_str()?;
    if !cfg.devices.iter().any(|d| d == device) {
        return None;
    }
    if !valid_command_id(object.get("id")?.as_str()?) {
        return None;
    }
    Some(IncomingCommand {
        agent: parts[1].into(),
        data,
    })
}
fn ack(data: &Value, status: &str, reason: &str) -> (String, Value) {
    (
        format!("agentdeck/{}/ack", data["device_id"].as_str().unwrap()),
        json!({"id":data["id"],"status":status,"reason":reason,"ts":unix_time()}),
    )
}

fn handle(db: &Connection, cfg: &Config, item: IncomingCommand) -> (String, Value) {
    let data = item.data;
    let now = unix_time();
    let Some(action) = data["action"].as_str() else {
        return ack(&data, "rejected", "expired_or_invalid");
    };
    let valid = ACTIONS.contains(&action)
        && data["ts"]
            .as_i64()
            .is_some_and(|v| (now - 15..=now + 5).contains(&v))
        && data["expires_at"]
            .as_i64()
            .is_some_and(|v| (now..=now + 30).contains(&v));
    if !valid {
        return ack(&data, "rejected", "expired_or_invalid");
    }
    let device = data["device_id"].as_str().unwrap();
    let id = data["id"].as_str().unwrap();
    let existing = db
        .query_row(
            "SELECT agent,action,status FROM commands WHERE device=?1 AND id=?2",
            params![device, id],
            |r| {
                Ok((
                    r.get::<_, String>(0)?,
                    r.get::<_, String>(1)?,
                    r.get::<_, String>(2)?,
                ))
            },
        )
        .ok();
    if let Some((old_agent, old_action, status)) = existing {
        return ack(
            &data,
            if old_agent == item.agent && old_action == action && status != "executing" {
                &status
            } else {
                "rejected"
            },
            "duplicate_or_unknown",
        );
    }
    let (state, _) = read_state(&cfg.state_dir, &item.agent, now);
    let online = state["online"].as_bool().unwrap_or(false);
    let argv = cfg.handlers.get(&item.agent).and_then(|m| m.get(action));
    if !online || argv.is_none() {
        return ack(
            &data,
            "rejected",
            if online {
                "unsupported_action"
            } else {
                "agent_offline"
            },
        );
    }
    db.execute(
        "INSERT INTO commands VALUES(?1,?2,?3,?4,'executing',?5)",
        params![device, id, item.agent, action, now],
    )
    .ok();
    let mut envelope = data.clone();
    envelope["agent"] = json!(item.agent);
    let bytes = serde_json::to_vec(&envelope).unwrap();
    let argv = argv.unwrap();
    let result = Command::new(&argv[0])
        .args(&argv[1..])
        .stdin(Stdio::piped())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
        .and_then(|mut child| {
            child.stdin.take().unwrap().write_all(&bytes)?;
            let deadline = Instant::now() + Duration::from_secs(5);
            loop {
                if let Some(status) = child.try_wait()? {
                    break Ok(status.success());
                }
                if Instant::now() >= deadline {
                    let _ = child.kill();
                    let _ = child.wait();
                    break Ok(false);
                }
                thread::sleep(Duration::from_millis(25));
            }
        })
        .unwrap_or(false);
    let status = if result { "completed" } else { "failed" };
    db.execute(
        "UPDATE commands SET status=?1 WHERE device=?2 AND id=?3",
        params![status, device, id],
    )
    .ok();
    db.execute("DELETE FROM commands WHERE ts < ?1", params![now - 86400])
        .ok();
    ack(&data, status, "")
}

pub fn run(cfg: Config, stop: Arc<AtomicBool>) -> Result<()> {
    std::fs::create_dir_all(&cfg.state_dir)?;
    let mut options = MqttOptions::new(&cfg.client_id, &cfg.host, cfg.port);
    options.set_keep_alive(Duration::from_secs(10));
    options.set_last_will(LastWill::new(
        "pc/status",
        serde_json::to_vec(&json!({"online":false,"ts":unix_time()}))?,
        QoS::AtLeastOnce,
        true,
    ));
    if let Ok(user) = std::env::var(&cfg.username_env) {
        options.set_credentials(user, std::env::var(&cfg.password_env).unwrap_or_default());
    }
    if let Some(ca) = &cfg.tls_ca {
        options.set_transport(Transport::tls(std::fs::read(ca)?, None, None));
    }
    let (client, mut connection) = Client::new(options, 32);
    let (command_tx, command_rx) = mpsc::sync_channel(16);
    let (out_tx, out_rx) = mpsc::sync_channel::<(String, Value)>(64);
    let connected = Arc::new(AtomicBool::new(false));
    let event_stop = stop.clone();
    let event_connected = connected.clone();
    let event_cfg = cfg.clone();
    let event_client = client.clone();
    let event = thread::spawn(move || {
        while !event_stop.load(Ordering::Relaxed) {
            match connection.recv_timeout(Duration::from_millis(200)) {
                Ok(Ok(Event::Incoming(Incoming::ConnAck(_)))) => {
                    for agent in AGENTS {
                        let _ = event_client
                            .subscribe(format!("agent/{agent}/command"), QoS::AtLeastOnce);
                    }
                    event_connected.store(true, Ordering::Relaxed);
                    println!("MQTT connected");
                }
                Ok(Ok(Event::Incoming(Incoming::Publish(p)))) => {
                    if let Some(c) = accepted(&p, &event_cfg) {
                        let _ = command_tx.try_send(c);
                    }
                }
                Ok(Ok(Event::Outgoing(Outgoing::Disconnect)))
                | Ok(Err(ConnectionError::RequestsDone)) => break,
                Ok(Err(e)) => {
                    event_connected.store(false, Ordering::Relaxed);
                    eprintln!("MQTT: {e}");
                    thread::sleep(Duration::from_secs(1));
                }
                Err(_) | Ok(Ok(_)) => {}
            }
        }
    });
    let worker_cfg = cfg.clone();
    let worker_stop = stop.clone();
    let worker = thread::spawn(move || {
        let db = Connection::open(worker_cfg.state_dir.join("commands.sqlite3")).unwrap();
        db.execute("CREATE TABLE IF NOT EXISTS commands(device TEXT,id TEXT,agent TEXT,action TEXT,status TEXT,ts INTEGER,PRIMARY KEY(device,id))",[]).unwrap();
        while !worker_stop.load(Ordering::Relaxed) {
            if let Ok(item) = command_rx.recv_timeout(Duration::from_millis(200)) {
                let _ = out_tx.try_send(handle(&db, &worker_cfg, item));
            }
        }
    });
    let usage = UsageCache::start(cfg.codex_usage_command.clone(), stop.clone());
    let mut metrics = PcMetrics::new();
    let mut next = Instant::now();
    while !stop.load(Ordering::Relaxed) {
        if connected.load(Ordering::Relaxed) && Instant::now() >= next {
            next = Instant::now() + Duration::from_secs(cfg.interval);
            let body = metrics.sample();
            let _ = client.publish(
                "pc/status",
                QoS::AtMostOnce,
                false,
                serde_json::to_vec(&body)?,
            );
            for agent in AGENTS {
                let (mut status, state_usage) = read_state(&cfg.state_dir, agent, unix_time());
                if agent == "codex"
                    && !status["online"].as_bool().unwrap_or(false)
                    && usage.reachable()
                {
                    status["online"] = json!(true);
                    status["working"] = json!(false);
                }
                let agent_usage = if agent == "codex" {
                    usage.sample().or(state_usage)
                } else {
                    state_usage
                };
                let usage_body =
                    agent_usage.unwrap_or_else(|| json!({"available":false,"ts":unix_time()}));
                let _ = client.publish(
                    format!("agent/{agent}/status"),
                    QoS::AtMostOnce,
                    false,
                    serde_json::to_vec(&status)?,
                );
                let _ = client.publish(
                    format!("agent/{agent}/usage"),
                    QoS::AtMostOnce,
                    false,
                    serde_json::to_vec(&usage_body)?,
                );
            }
        }
        while let Ok((topic, body)) = out_rx.try_recv() {
            if unix_time() - body["ts"].as_i64().unwrap_or(0) <= 15 {
                let _ = client.publish(topic, QoS::AtMostOnce, false, serde_json::to_vec(&body)?);
            }
        }
        thread::sleep(Duration::from_millis(50));
    }
    for agent in AGENTS {
        let _ = client.publish(
            format!("agent/{agent}/status"),
            QoS::AtMostOnce,
            false,
            serde_json::to_vec(&json!({"online":false,"working":false,"ts":unix_time()}))?,
        );
    }
    let _ = client.publish(
        "pc/status",
        QoS::AtLeastOnce,
        true,
        serde_json::to_vec(&json!({"online":false,"ts":unix_time()}))?,
    );
    let _ = client.disconnect();
    let _ = worker.join();
    let _ = event.join();
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn retained_and_unauthorized_rejected() {
        let mut cfg = Config::default();
        cfg.host = "x".into();
        cfg.devices = vec!["deck".into()];
        let p = Publish::new(
            "agent/codex/command",
            QoS::AtLeastOnce,
            json!({"device_id":"deck","id":"x"}).to_string(),
        );
        assert!(accepted(&p, &cfg).is_some());
        let mut retained = p.clone();
        retained.retain = true;
        assert!(accepted(&retained, &cfg).is_none());
    }
}
